prusti-interface/src/environment/mod.rs

```diff
@@ -157,7 +162,13 @@ impl<'tcx> Environment<'tcx> {
         let tcx = self.tcx;
         let mut visitor = CollectPrustiSpecVisitor::new(self);
         tcx.hir().krate().visit_all_item_likes(&mut visitor);
-        visitor.get_annotated_procedures()
+
+        let mut cl_visitor = CollectClosureDefsVisitor::new(self);
```

Comment:
Well, yes, how else are you going to do it if a) closures are encoded as functions in the MIR and b) the MIR is the basis for verification in Prusti?

This doesn't make the approach less powerful, however, only a little more verbose. Consider this (using informal syntax):
```Rust
let x = 42;
let cl = /* ensures: result == 42 */ || x;
```
This won't verify, because we do indeed lose information about the captured state. However, we can define a view:
```Rust
let x = 42;
let cl =
    // view: capt_x: i32 = x
    // ensures: result == capt_x
    || x;
assert_eq!(cl(), 42);
```
This will verify (once everything is implemented, I mean), because we can prove the specification separately _and_ use the knowledge about the captured state available at the instantiation-site to deduce the actual return value. I suppose you could even give an invariant:
```Rust
let x = 42;
let cl =
    // view: capt_x: i32 = x
    // invariant: capt_x == 42
    // ensures: result == 42
    || x;
```
Then we can prove the original postcondition, although of course the only interesting aspect about this "invariant" is that it is established; we get preservation for free.

Reply:
The implementation of closures in Nagini used to essentially inline the closure bodies into the parent function so that they could take the captured state into account. However, your approach with views also looks reasonable. Thanks for explaining it!